storagenode/pieces/readwrite.go

```diff
@@ -79,16 +193,80 @@ func NewReader(blob storage.BlobReader) (*Reader, error) {
+	if formatVersion >= storage.FormatV1 {
+		if size < V1PieceHeaderReservedArea {
+			return nil, Error.New("invalid piece file for storage format version %d: too small for header (%d < %d)", formatVersion, size, V1PieceHeaderReservedArea)
+		}
+		size -= V1PieceHeaderReservedArea
+	}
 
-	reader := &Reader{}
-	reader.blob = blob
-	reader.size = size
-
+	reader := &Reader{
+		formatVersion: formatVersion,
+		blob:          blob,
+		size:          size,
+	}
 	return reader, nil
 }
 
+// GetStorageFormatVersion returns the storage format version of the piece being read.
+func (r *Reader) GetStorageFormatVersion() storage.FormatVersion {
+	return r.formatVersion
+}
+
+// GetPieceHeader reads, unmarshals, and returns the piece header. It may only be called
+// before any Read() calls. (Retrieving the header at any time could be supported, but for
+// the sake of performance we need to understand why and how often that would happen.)
+func (r *Reader) GetPieceHeader() (*pb.PieceHeader, error) {
+	if r.formatVersion < storage.FormatV1 {
+		return nil, Error.New("Can't get piece header from storage format V0 reader")
+	}
+	if r.pos != 0 {
+		return nil, Error.New("GetPieceHeader called when not at the beginning of the blob stream")
+	}
+	// We need to read the size of the serialized header protobuf before we read the header
+	// itself. The headers aren't a constant size, although V1PieceHeaderReservedArea is
+	// constant. Without this marker, we wouldn't have any way to know how much of the
+	// reserved header area is supposed to make up the serialized header protobuf.
+	var headerBytes [V1PieceHeaderReservedArea]byte
+	framingBytes := headerBytes[:v1PieceHeaderFramingSize]
+	n, err := r.blob.Read(framingBytes)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+	if n != v1PieceHeaderFramingSize {
+		return nil, Error.New("Could not read whole PieceHeader framing field")
+	}
+	r.pos += int64(n)
+	headerSize := binary.BigEndian.Uint16(framingBytes)
+	if headerSize > (V1PieceHeaderReservedArea - v1PieceHeaderFramingSize) {
+		return nil, Error.New("PieceHeader framing field claims impossible size of %d bytes", headerSize)
+	}
+
+	// Now we can read the actual serialized header.
+	pieceHeaderBytes := headerBytes[v1PieceHeaderFramingSize : v1PieceHeaderFramingSize+headerSize]
+	n, err = r.blob.Read(pieceHeaderBytes)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+	r.pos += int64(n)
+
+	// Deserialize and return.
+	header := &pb.PieceHeader{}
+	if err := proto.Unmarshal(pieceHeaderBytes, header); err != nil {
+		return nil, Error.New("piece header: %v", err)
+	}
+	return header, nil
+}
+
 // Read reads data from the underlying blob, buffering as necessary.
 func (r *Reader) Read(data []byte) (int, error) {
+	if r.formatVersion >= storage.FormatV1 && r.pos < V1PieceHeaderReservedArea {
+		// should only be necessary once per reader. or zero times, if GetPieceHeader is used
+		if _, err := r.Seek(0, io.SeekStart); err != nil {
```

Comment:
Can you help me understand this? I thought that we would want to do

`r.Seek(V1PieceHeaderReservedArea, io.SeekStart)`

So that we would start reading _after_ the header reserved area? It appears here if we haven't read past the area, we seek back to the very beginning, which would simply read the reserved area?

Reply:
Neeeevermind. I should have read the next function first.

Reply:
Yeah, this is a little confusing. Maybe it should work in terms of `r.blob.Seek` instead?

Reply:
I'm good either way. It's confusing when used internally, but externally to have the Seek function work only on the actual piece data makes more sense.